### ISSUE TITLE
Improve documentation on rebuilding maps

### ISSUE THREAD

**Issue by andreasspeck** [manifestinteractive/jqvmap#182]
The documentation on building your own maps seems to be outdated - or the reference to the jvectormap documentation seems to be confusing, as jvectormap has moved on. converter.py of release 2.0.4 of jvectormap does not match the bits of documentation given on jqvmap.

I installed converter.py version-1.2.2 and 1.1.1, and found some recipe/documentation burried in old versions of the website on rebuilding the world map.

There are some fixes needed:
- both, converter.py 1.1.1 and 1.2.2 miss a line at the very beginning of the script (the import section): import shapely.wkb . This is crucial, as otherwise converter.py will crash
- nevertheless, rebuilding the world map with the latest data from naturalearthdata.com fails:

python2.7 ../converter-1.2.2/converter.py --width 950 --where "iso_a2 != 'AQ'" --projection "merc" --input_file_encoding "utf-8" --longitude0 11.5 --country_code_index 26 --country_name_index 11 --simplify_tolerance 1000 --buffer_distance -0.3 --name "world-en" ../ne_110m_admin_0_countries.shp world-en.js

Self-intersection at or near point 18766847.944786005 -1870058.4770726727
Self-intersection at or near point 18766847.944786005 11350893.896457316

This might be a problem with the source data... Any suggestion on how to solve it, or should I ask at jvectormap (although, as this concerns an old release of jvectormap, this might not be useful either).

I am also not sure the resulting world-en.js really has the same structure as jquery.vmap.world.js - I notice an additional inset and bbox. 

Any help is appreciated (and would possibly solve the issue of people asking to add a specific country).

**Comment by manifestinteractive** [comment]
@andreasspeck I personally cannot help with this as I have never actually gotten this to work.  As mentioned in the CONTRIBUTING docs, I personally do not contribute to anything that has to do with map creation/manipulation and purposefully link to the place that had docs for this.

If someone can get a working example up and running, document the heck out of it for everyone, then I will be more than happy to reintegrate it into this project.

**Comment by manifestinteractive** [comment]
@andreasspeck I've decided that since there are enough people asking about this, I should revisit this again personally.  I was honestly hoping on of the thousand or so people that starred this project might be a little more familiar with Python, which I am not :(  But if I can get it working, I will post the instructions as an update.

**Comment by manifestinteractive** [comment]
@andreasspeck OK, so five hours later I have this working 100%.  I have created a new branch at https://github.com/manifestinteractive/jqvmap/tree/map-creator for you to try out.  If this works for you, let me know and I will merge it into master.

Please feel free to fork the branch and make modifications to the README in the ./create folder if you find the instructions unclear.

Hope this is helpful.  Not being able to get maps working has always been a pain in my butt.

**Comment by andreasspeck** [comment]
Thanks, I tried this with the world map from NaturalEarthData.

My world.json file is:

[
  {
    "name": "read_data",
    "file_name": "./source/ne_110m_admin_0_countries_lakes.shp"
  },
  { 
    "name": "write_data",
    "format": "jqvmap",
    "file_name": "./output/jquery.vmap.world-orig.js",
    "params": {
      "code_field": "iso_a2",
      "name_field": "name_long",
      "name": "world_en_orig"
    }
  }
]

(I think iso_a2 has the two letter country code).

I get the following error:

ERROR 1: Full reprojection failed, but partial is possible if you define OGR_ENABLE_PARTIAL_REPROJECTION configuration option to TRUE

I haven't tried what it says there.

Also, the resulting map js file has an error. In the js file, the projection data starts with "PATHS", while I think jqvmap requires it to be "PATHES". When you adjust that, the map at least displays, although it is slightly broken.

**Comment by andreasspeck** [comment]
OK, that error goes away when I do:

export OGR_ENABLE_PARTIAL_REPROJECTION=TRUE

before running python processor.py config/world.json 

I then get another error/warning:
Self-intersection at or near point -15921827.784687197 -19625815.332566325

But it creates the map - which still includes Antartica, which I need to exclude somehow.

**Comment by andreasspeck** [comment]
Some more experiments, looking at the documentation at http://jvectormap.com/maps/world/world/

I added these two lines to world.json:

"filter": "iso_a2 != \"AQ\"",
"longitude0": 11.5

Although I still get:

Self-intersection at or near point 18766847.944786005 -1870058.4770726701
Self-intersection at or near point 18766847.944786005 11350893.896457318

I now get a functioning map.
(I changed in converter.py and processor.py the following line:
  def getJSCode(self):
    map = {"paths": self.paths, ....

to;
    def getJSCode(self):
    map = {"pathes": self.paths,
)

This now creates a functioning map.

One note though: this map now has the country codes in upper case, while the map distributed previously with JQVMap had them in lower case. This makes a difference.

**Comment by manifestinteractive** [comment]
@andreasspeck I will make the change in the JQVMap to actually use `paths` rather than `pathes`.  This will allow people to use maps between the projects.

In regards to the uppercase vs lowercase, technically this is just going to be whatever was in the Shapefiles themselves.  It sounds like you might be suggesting we use lowercase.  I actually agree it might be a good idea to do this in python and force lowercase for the codes.  Otherwise this might be weird for people looking for codes that do not actually match.

The only issue I can think of is in your example `iso_a2 != \"AQ\"` we'd have to figure out how to handle case sensitivity there too, should people put in lowercase ... I wonder if that works without issues as is.

Something we should probably check out, as I see that as being a potential issue in the future.

Glad you got this working, and thanks for being the first to test it :)

I will work on cleaning up the loose ends.  In the meantime, if there was anything in the documentation that you found could be written better, please let me know.  A PR on this branch would be even better :)

**Comment by andreasspeck** [comment]
Thanks for all your work, as I'm just working on a project using JQVMap. 

I need to figure out how to do a PR - I can think of some improvements in the documentation for Linux users (I think most of the dependencies can be installed from packages in Debian, besides booleano).

The issue of country codes in lower or upper case is not really important. I only mentioned it because I think the maps that are distributed with jqvmap have them in lowercase, and if someone has written their application around that, that might require them to be in lowercase (for example WorldTagMap for Joomla - which is how I found jqvmap), otherwise their applications will stop working.
The best would be to have that as a configurable option.

**Comment by manifestinteractive** [comment]
It's all good.  I just updated the create/README and updated a few things.

1.  renamed processor.py to jqvmap.py
2.  removed unused python files
3.  reworked jqvmap.py to use lowercase codes consistently
4.  updated core dist files to use `paths` rather than `pathes` as this will allow for use of maps between the vector map libraries
5.  updated existing map files to use `paths` rather than `pathes`

Bottom line is I want people to be able to share these maps with other libraries so this minor update will allow for that.  I will do a few more tests on this and release it if it all looks good.

Thanks again for your help testing this.

**Comment by andreasspeck** [comment]
Thanks a lot for your work ;-)

The change related to paths rather than pathes is good - as that makes the maps compatible with jvectormap at least. I'll get your latest version than for the project I am working on.
The example maps you can download from jvectormap all have UPPERCASE country codes - as that's what's in the DBF files from NaturalEarth. 
The best would be if that could be case insensitive - for example that jqvmap would always pass them in lower case to the function, no matter what case they are in in the map.js file. That means as a user who just wants to reuse existing maps you do not have to worry about that.
It's probably no big deal.

**Comment by andreasspeck** [comment]
I guess what I mean is:
- for example, when I initialise a map and want to define the colours of certain regions, I pass the region/country codes - and these are case sensitive! Is that necessary? Do I need to be aware whether the country codes in the map files are lowercase or uppercase? I don't see a case where you would have a need for case sensitive country codes, so could we do away with that, and deal with "ES" the same way as with "es"?

**Comment by manifestinteractive** [comment]
That's a good idea.  It should not need to be case sensitive as the code is almost always the ISO code.  This is normally written in upper case, but JS should not care about the case.  I will update the main DIST files to use lowercase even if you provide uppercase codes.

**Comment by andreasspeck** [comment]
Well, if you can make it even more flexible, so that it doesn't matter what's in the map file (the map files you can download from jvectormap have the country codes in uppercade), that would be even better. That would improve compatibility.
Otherwise, we would need a script that converts map files with uppercase country codes into map files with lowercase country codes.

**Comment by manifestinteractive** [comment]
it should not matter what is in the map file, but I think if I am reworking the generator, which I am, it should be consistent if nothing else.

In regards to the plugin though, I will look for any conditions where you are looking up something based off a code, and add a condition to check for both upper and lower case versions.  So you can use either and it will still work as a selector.

**Comment by andreasspeck** [comment]
Thanks - that would be perfect, as it would save me the pain to deal with uppercase and lowercase.

**Comment by manifestinteractive** [comment]
Can you let me know which methods of JQVMap you are using to target specific codes?  I want to make sure I don't miss something you already know you want to do.

**Comment by manifestinteractive** [comment]
This is now merged into master and a tag branch was created for v1.3.0.  See CHANGELOG.md for important notes.
